trin-execution/src/walkers/trie_walker.rs

```diff
@@ -0,0 +1,378 @@
+            trie: Arc::new(trie),
+            stack: vec![root_proof],
+        })
+    }
+
+    fn process_node(
+        &mut self,
+        node: Node,
+        partial_proof: Vec<Bytes>,
+        path: Vec<u8>,
+    ) -> anyhow::Result<()> {
+        // We only need to process hash nodes, because if the node isn't a hash node the leaf is
+        // already embedded in the proof
+        if let Node::Hash(hash) = node {
+            let value_result = self
+                .trie
+                .get(hash.hash.as_slice())
+                .expect("Failed to read node from the database");
+
+            let value = match value_result {
+                Some(value) => value,
+                None => {
+                    // If we are walking a partial trie, some nodes won't be available in the
+                    // database
+                    if self.is_partial_trie {
+                        return Ok(());
+                    }
+                    return Err(anyhow::anyhow!("Node not found in the database"));
+                }
+            };
+            let decoded_node = decode_node(&mut value.as_ref())?;
+            match decoded_node {
+                Node::Leaf(_) | Node::Extension(_) | Node::Branch(_) => {
+                    self.stack.push(TrieProof {
+                        path,
+                        proof: [partial_proof, vec![value.into()]].concat(),
+                    });
+                }
+                // can't be a hash node because we just decoded it
+                Node::Empty | Node::Hash(_) => (),
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<TrieDB: DB> Iterator for TrieWalker<TrieDB> {
+    type Item = TrieProof;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let next_proof = match self.stack.pop() {
+            Some(next_proof) => next_proof,
+            None => return None,
+        };
+
+        let TrieProof { path, proof } = &next_proof;
+        let last_node = proof.last().expect("Proof is empty");
+        let decoded_last_node =
+            decode_node(&mut last_node.as_ref()).expect("Failed to decode node");
+
+        // Process any children of the node
+        match decoded_last_node {
+            Node::Extension(extension) => {
+                let extension = extension.read().expect("Extension node must be readable");
+                self.process_node(
+                    extension.node.clone(),
+                    proof.clone(),
+                    [
+                        path.as_slice(),
+                        extension.prefix.get_data().to_vec().as_slice(),
+                    ]
+                    .concat(),
+                )
+                .expect("Failed to process node");
+            }
+            Node::Branch(branch) => {
+                let branch = branch.read().expect("Branch node must be readable");
+
+                // We don't need to check the branches value as it is already encoded in the proof
```

Comment:
nit: I think this comment is not needed. We are iterating Trie nodes, not Trie values.